File: 排序/148. 排序链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var sortList = function(head) {
    // 如果 head 为空值的话直接返回
    if(head == null) return head;
    // 设置l 最小值， r 最大值， mid 基准值;
    let l = head.val, r = head.val, mid;
    // p， q 操作指针， 分支链表h1 ,h2
    let p = head , q , h1 = null, h2 = null;
    // 找到l ,r 链表中的最大值，最小值计算基准值
    while(p) {
        l = Math.min(p.val,l);
        r = Math.max(p.val,r);
        p = p.next;
    }
    // 如果最大值和最小值相等说明 只有一个值或者已经有序
    if(l == r) return head;
    // mid 基准值
    mid = (l + r) / 2;
    // 分隔链表为较小子链表和较大子链表
    p = head;
    // 将链表节点一次向子表头部插入
    while(p){
        // 记录下个节点
        q = p.next;
        if(p.val <= mid){
            p.next = h1;
            h1 = p;
        } else {
            p.next = h2;
            h2 = p;
        }
        // 移动操作指针
        p = q;
    }
    // 递归较小子链表
    h1 = sortList(h1);
    // 递归较大子链表
    h2 = sortList(h2);
    // 操作较小子链表，
    p = h1;
    // 将p 指针指向较小子链表的尾部
    while(p.next) p = p.next;
    // 连接两链表
    p.next = h2;
    // 返回表头h1
    return h1;
};

File: Python/148. 排序链表.py
# -*- codeing = utf-8 -*-
# @Time : 2023-8-23 22:34
# @Author : 刘永奇
# @File : 148. 排序链表.py
# @Software : PyCharm
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.sortFunc(head, None)

    def sortFunc(self, left, right):
        if not left:
            return left
        if left.next == right:
            left.next = None
            return left
        slow = fast = left
        while fast != right:
            slow = slow.next
            fast = fast.next
            if fast != right:
                fast = fast.next
        mid = slow
        return self.merge(self.sortFunc(left, mid), self.sortFunc(mid, right))

    def merge(self, L1, L2):
        T = ListNode(0)
        temp, temp1, temp2 = T, L1, L2
        while temp1 and temp2:
            if temp1.val < temp2.val:
                temp.next = temp1
                temp1 = temp1.next
            else:
                temp.next = temp2
                temp2 = temp2.next
            temp = temp.next
        if temp1:
            temp.next = temp1
        elif temp2:
            temp.next = temp2
        return T.next
